ethpm_cli/scraper.py

```diff
@@ -0,0 +1,215 @@
+            raise BlockNotFoundError(
+                f"Block number: {start_block} not available on provided web3 "
+                f"instance with latest block number of {w3.eth.blockNumber}."
+            )
+
+        all_scraped_blocks = get_scraped_blocks(chain_data_path)
+        if start_block in all_scraped_blocks:
+            raise BlockAlreadyScrapedError(
+                f"Skipping block #{start_block}. Already processed."
+            )
+        return start_block
+    else:
+        ethpm_dir.mkdir()
+        # make base dir for all ipfs assets
+        (ethpm_dir / "Qm").mkdir()
+        chain_data_path = ethpm_dir / "chain_data.json"
+        chain_data_path.touch()
+        init_json = {
+            "chain_id": w3.eth.chainId,
+            "registry_addresses": [],
+            "scraped_blocks": [{"min": "0", "max": "0"}],
+        }
+        chain_data_path.write_text(f"{json.dumps(init_json, indent=4)}\n")
+        return 1
+
+
+def update_chain_data(
+    chain_data_path: Path, block_number: int, manifests: Dict[Address, Dict[str, str]]
+) -> None:
+    chain_data = json.loads(chain_data_path.read_text())
+
+    new_addrs = list(manifests.keys())
+    updated_addrs = list(sorted(set(chain_data["registry_addresses"] + new_addrs)))
+
+    all_scraped_blocks = get_scraped_blocks(chain_data_path)
+    updated_blocks = blocks_to_ranges(set(all_scraped_blocks + [block_number]))
+
+    chain_data_updated_blocks = assoc(chain_data, "scraped_blocks", updated_blocks)
+    chain_data_updated_addrs = assoc(
+        chain_data_updated_blocks, "registry_addresses", updated_addrs
+    )
+    chain_data_path.write_text(f"{json.dumps(chain_data_updated_addrs, indent=4)}\n")
+
+
+@to_list
+def blocks_to_ranges(blocks_list: List[int]) -> Iterable[Dict[str, str]]:
+    """
+    Takes a list of block numbers, and returns them grouped into min/max ranges.
+    """
+    for a, b in itertools.groupby(enumerate(blocks_list), lambda x: x[0] - x[1]):
+        b = list(b)  # type: ignore
+        yield {"min": str(b[0][1]), "max": str(b[-1][1])}  # type: ignore
+
+
+def get_scraped_blocks(chain_data_path: Path) -> List[int]:
+    scraped_blocks = json.loads(chain_data_path.read_text())["scraped_blocks"]
+    scraped_ranges = [
+        list(range(int(rnge["min"]), int(rnge["max"]) + 1)) for rnge in scraped_blocks
+    ]
+    return flatten(scraped_ranges)
+
+
+def write_ipfs_uris_to_disk(
+    ethpm_dir: Path, manifests: Dict[Address, Dict[str, str]]
+) -> None:
+    ipfs_dir = ethpm_dir / "Qm"
+    all_manifest_uris = [
+        version_release_data["manifestURI"]
+        for version_release_data in manifests.values()
+        if is_supported_content_addressed_uri(version_release_data["manifestURI"])
+    ]
+    base_ipfs_uris = [uri for uri in all_manifest_uris if is_ipfs_uri(uri)]
+    nested_ipfs_uris = [pluck_ipfs_uris_from_manifest(uri) for uri in all_manifest_uris]
+    all_ipfs_uris = set(flatten(nested_ipfs_uris) + base_ipfs_uris)
```

Comment:
This might be more appropriate done as `set(flatten(nested_ipfs_uris)).update(*base_ipfs_uris)`